common/lib/core-interfaces/src/provider.ts

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * @internal
+ * This utility type is primarily meant for internal use by @see Provider
+ * Produces a valid Provider key given a type and a property.
+ * A valid Provider key is a property that exists on the incoming type
+ * as well as on the type of the property itself. For example, IProvideFoo.IFoo.IFoo
+ * This aligns with the provider pattern expected to be used with all Providers.
+ * For example:
+ * ```
+ * interface IProvideFoo{
+ *  IFoo: IFoo
+ * }
+ * interface IFoo extends IProvideFoo{
+ *  foobar();
+ * }
+ * ```
+ * This pattern enables discovery, and delegation in a standard way which is central
+ * to Provider pattern
+ */
+ export type ProviderPropertyKeys<T, TProp extends keyof T = keyof T> =
+    string extends TProp ? never : number extends TProp? never : // exclude indexers [key:string |number]: any
+    TProp extends keyof T[TProp] // TProp is a property of T, T[TProp] and, T[TProp][TProp]
+        ? TProp extends keyof T[TProp][TProp] // ex; IProvideFoo.IFoo.IFoo.IFoo
+            ? TProp
+            :never
+        : never;
+
+/**
+ * This utility type take interface(s) that follow the provider pattern, and produces
+ * a new type that can be used for inspection and discovery of those interfaces.
+ *
+ * It is meant to be used with types that are known to implement the provider pattern.
+ * A common way to specify a type implements the provider pattern is to expose it as a
+ * Provider without a generic argument.
+ *
+ * For example, if we have an interface like below
+ * ```
+ * interface IProvideFoo{
+ *  IFoo: IFoo
+ * }
+ * interface IFoo extends IProvideFoo{
+ *  foobar();
+ * }
+ * ```
+ *
+ * and a function that returns a Provider. You would do the following
+ *
+ * `const maybeFoo : Provider<IFoo> = getUnknown()`;
```

Comment:
Trying this out in TS Playground. I have to do it this way:

```suggestion
 * `const maybeFoo = getUnknown() as Provider<IFoo>;`
```

Reply:
can you share what you tried? there are test below that validate exactly this

Reply:
[TS Playground](https://www.typescriptlang.org/play?ts=4.1.5#code/MYewdgzgLgBANiA5jAvDUkRwKYDoGIDcAUMVAJ4AO2MACgE4gBuAlgCbb0MjX0UDS2chAA8AFQA0MMd0oxsADyjYwbCDADWQkADNpqTdr1iAfKmIxLMaPRZhki5avUzGcgPwww2JpxgAuLwBXAFsAIz9HFTVpWU9vX3oAmAB6FPkFYDggjhg7DgVOdQBtLXJ-GztkAB8wUIj6AF1AgEMwcgsrVx4Mpxiy3Wli7spG1PSRvPUWmEo3TgoYQckhkbG2tikxYdlGnbdGzqsrT0mo50NyZf2ePbXxjMIYAEluVg4AMRAQXGevn7+31+-yOxzBp1koLBVkCCU4UJhXh8nBIZCoNDe7E44gMQTAGjAIAA7mAzGhaC0+CwWnARLQWMANOIpJiOFx5nxyIJhOITHzURRqHRGO9ONzRKYDAM9KzsaZUTo8cAoCxwDAoPRyCwoAAKACOgTxBOJYAAlDAAN6gjDQGAhFrkCIAIUpgVl9BEzxd9DJMD1JGOBB19sd2G9v297lw5FNJAAvqQsi0IOpvXkQpQcCEVFB1F7KZbQeVrBqqgYAEQAQQAyh8ACLlgNWfP0QItgxQAAWLAg8dIdmU9B0LWANFeIqxaatxxbbe9feIA84w9HLzT5xi4+Yk4L06sxcq9gXGq1uu8RJg3p1ptjxCAA)

Reply:
It appears that `Provider` is equivalent to `{}` not `unknown`.  Try adding this line in the TS Playgroudn link above:

`type ProviderUnknown = Provider<unknown>`

And hover over `ProviderUnknown`.

Reply:
Provider is meant to be equivalent to `{}` is not supposed to be equivalent to `unknown`. i don't see a problem with this, except maybe the comment are not clear enough